core/lib/types/src/contract_verification/contract_identifier.rs

```diff
@@ -0,0 +1,388 @@
+use serde::{Deserialize, Serialize};
+
+use crate::{bytecode::BytecodeMarker, web3::keccak256, H256};
+
+/// An identifier of the contract bytecode.
+/// This identifier can be used to detect different contracts that share the same sources,
+/// even if they differ in bytecode verbatim (e.g. if the contract metadata is different).
+///
+/// Identifier depends on the marker of the bytecode of the contract.
+/// This might be important, since the metadata can be different for EVM and EraVM,
+/// e.g. `zksolc` [supports][zksolc_keccak] keccak256 hash of the metadata as an alternative to CBOR.
+///
+/// [zksolc_keccak]: https://matter-labs.github.io/era-compiler-solidity/latest/02-command-line-interface.html#--metadata-hash
+// Note: there are missing opportunities here, e.g. Etherscan is able to detect the contracts
+// that differ in creation bytecode and/or constructor arguments (for partial match). This is
+// less relevant for ZKsync, since there is no concept of creation bytecode there; although
+// this may become needed if we will extend the EVM support.
+#[derive(Debug, Clone, Copy)]
+pub struct ContractIdentifier {
+    /// Marker of the bytecode of the contract.
+    pub bytecode_marker: BytecodeMarker,
+    /// SHA3 (keccak256) hash of the full contract bytecode.
```

Comment:
AFAIU, SHA3-256 and keccak256 differ by the finalization padding (in other words, they are 2 different hash functions using the same general construction). I.e., to avoid confusion, it makes sense to mention keccak256 only.